packages/block-editor/src/components/inserter/search-items.js

```diff
@@ -44,94 +43,98 @@ const removeMatchingTerms = ( unmatchedTerms, unprocessedTerms ) => {
 	);
 };
 
+export const searchBlockItems = (
+	items,
+	categories,
+	collections,
+	searchTerm
+) => {
+	const normalizedSearchTerms = normalizeSearchTerm( searchTerm );
+	if ( normalizedSearchTerms.length === 0 ) {
+		return items;
+	}
+
+	return searchItems( items, searchTerm, {
+		getCategory: ( item ) =>
+			find( categories, { slug: item.category } )?.title,
+		getCollection: ( item ) =>
+			collections[ item.name.split( '/' )[ 0 ] ]?.title,
+		getVariations: ( item ) =>
+			( item.variations || [] ).map( ( variation ) => variation.title ),
+	} ).map( ( item ) => {
+		if ( isEmpty( item.variations ) ) {
+			return item;
+		}
+
+		const matchedVariations = item.variations.filter( ( variation ) => {
+			return (
+				intersectionWith(
+					normalizedSearchTerms,
+					normalizeSearchTerm( variation.title ),
+					( termToMatch, labelTerm ) =>
+						labelTerm.includes( termToMatch )
+				).length > 0
+			);
+		} );
+		// When no patterns matched, fallback to all variations.
+		if ( isEmpty( matchedVariations ) ) {
+			return item;
+		}
+
+		return {
+			...item,
+			variations: matchedVariations,
+		};
+	} );
+};
+
 /**
  * Filters an item list given a search term.
  *
  * @param {Array} items       Item list
- * @param {Array} categories  Available categories.
- * @param {Array} collections Available collections.
  * @param {string} searchTerm Search term.
- *
- * @return {Array}             Filtered item list.
+ * @param {Object} config     Search Config.
+ * @return {Array}            Filtered item list.
  */
-export const searchItems = ( items, categories, collections, searchTerm ) => {
+export const searchItems = ( items, searchTerm, config = {} ) => {
 	const normalizedSearchTerms = normalizeSearchTerm( searchTerm );
-
 	if ( normalizedSearchTerms.length === 0 ) {
 		return items;
 	}
 
-	return items
-		.filter(
-			( { name, title, category, keywords = [], variations = [] } ) => {
-				let unmatchedTerms = removeMatchingTerms(
-					normalizedSearchTerms,
-					title
-				);
-
-				if ( unmatchedTerms.length === 0 ) {
-					return true;
-				}
-
-				unmatchedTerms = removeMatchingTerms(
-					unmatchedTerms,
-					keywords.join( ' ' )
-				);
-
-				if ( unmatchedTerms.length === 0 ) {
-					return true;
-				}
-
-				unmatchedTerms = removeMatchingTerms(
-					unmatchedTerms,
-					get( find( categories, { slug: category } ), [ 'title' ] )
-				);
-
-				const itemCollection = collections[ name.split( '/' )[ 0 ] ];
-				if ( itemCollection ) {
-					unmatchedTerms = removeMatchingTerms(
-						unmatchedTerms,
-						itemCollection.title
-					);
-				}
-
-				if ( unmatchedTerms.length === 0 ) {
-					return true;
-				}
-
-				unmatchedTerms = removeMatchingTerms(
-					unmatchedTerms,
-					variations
-						.map( ( variation ) => variation.title )
-						.join( ' ' )
-				);
-
-				return unmatchedTerms.length === 0;
-			}
```

Comment:
Unless  I'm wrong here @gziolo it seems these different steps all do the same thing, so I just merged all terms in a single string of terms and call `removeMatchingTerms` just once.

Reply:
Yes, you nailed it. The previous implementation could terminate earlier when all matches were found but readability suffered because of that. All unit tests pass so it has to be good :)

Reply:
There are also e2e tests that I forgot to mention:
https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/specs/editor/plugins/block-variations.js

It gives me 99% confidence that refactoring is correct.